File: Ternary_Search/Ternary_Search.kt
    import java.util.*
    import kotlin.collections.ArrayList

    //Ternary Search begins
    public fun ternarySearch(l:Int, r:Int, key:Int, ar:ArrayList<Int>): Int {
        if(r>=1){
            var mid1 = l + (r - l) / 3
            var mid2 = r - (r - l) / 3

    //Searching for the key element
        if (ar[mid1] == key) {
            return mid1
        }
        
        if (ar[mid2] == key) {
            return mid2
        }
        
        if (key < ar[mid1]) {
            return ternarySearch(l, mid1 - 1, key, ar)
        }
        
        else if (key > ar[mid2]) {
            return ternarySearch(mid2 + 1, r, key, ar)
        }
        
        else {
            return ternarySearch(mid1 + 1, mid2 - 1, key, ar)
        }
    }
    return -1;
}

    //Main function
    fun main() {
        var l:Int = 0
        var r:Int
        var p:Int
        var key:Int
        var scan1: Scanner = Scanner(System.`in`)
    
    //var ar = arrayOf<Int>(1,2,3,4,5,6,7,8,9,10)
    var ar = ArrayList<Int>()

    //Taking input from user
    println("Enter the number of elements: ")
    var n:Int = scan1.nextInt()
    var temp:Int
        for (i in 1..n) {
            temp = scan1.nextInt()
            ar.add(temp)
         }

    //Sorting the array
    println("Sorting the array in ascending order...")
    ar.sort()
    r = ar.count()
    println("Enter the key element to search for:")
    key = scan1.nextInt()
    p = ternarySearch(l, r, key, ar)
    //Printing the key element with its position
    print("Index of $key is $p \n")
    
    
    /*Sample Input and Output
    Enter the number of elements :
    6
    5
    4
    3
    2
    1
    Sorting the array in ascending order...
    Enter the key element to search for:
    5
    Index of 5 is 4
    */

File: Ternary_Search/Ternary_Search.cs
using System;

public class ternarysrc {

    // Function to perform Ternary Search
    static int ternarySearch(int l, int r, int key, int[] ar)
    {
        while (r >= l)
        {

            // Finding the mid1 and mid2
            int mid1 = l + (r - l) / 3;
            int mid2 = r - (r - l) / 3;

            // To check if key is present at any mid
            if (ar[mid1] == key)
            {
                return mid1;
            }
            if (ar[mid2] == key)
            {
                return mid2;
            }

            // Checking region where KEY might be present
            if (key < ar[mid1])
            {
                r = mid1 - 1;
            }
            else if (key > ar[mid2])
            {
                l = mid2 + 1;
            }
            else
            {
                l = mid1 + 1;
                r = mid2 - 1;
            }
        }

        // If the KEY is not found
        return -1;
    }

    // Main Function to Execute
    public static void Main(String[] args)
    {
        int l, r, p, i, key = 0, len = 0;

        // Taking the input of Array Length from user
        Console.WriteLine("Enter Length of Array : ");
        len = Convert.ToInt32(Console.ReadLine());

        int[] ar = new int[len];

        Console.WriteLine("Enter Array Elements in Ascending Order : ");

        for(i = 0; i < len; i++ )       // Array input
        {
            ar[i] = Convert.ToInt32(Console.ReadLine());
        }

        // Starting index
        l = 0;

        // Ending Index
        r = len - 1;

        Console.WriteLine("\nEnter the Number to be Searched : ");
        // KEY to be searched in the array
        key = Convert.ToInt32(Console.ReadLine());

        // Searching the KEY using ternarySearch function
        p = ternarySearch(l, r, key, ar);


        // Checking if the KEY is present or not
        if(p == -1)
            Console.WriteLine("\nNumber Not Found");
        else
            Console.WriteLine("\n The Index of " + key + " is " + p);
        Console.ReadKey();


    }
}

// Sample Input/Output:
/*
    Enter Length of Array : 5
    Enter Array Elements in Ascending Order : 
    10
    20
    30
    40
    50

    Enter the Number to be Searched : 
    50

    The Index of 50 is 4

*/
